cmd/dagger/functions.go

```diff
@@ -694,6 +702,207 @@ func handleResponse(returnType *modTypeDef, response any, o, e io.Writer) error
 	return err
 }
 
+func handleChangesetResponse(ctx context.Context, dag *dagger.Client, response any) (rerr error) {
+	var changesetID string
+	switch v := response.(type) {
+	case string:
+		changesetID = v
+	default:
+		return fmt.Errorf("unexpected response type for changeset: %T", v)
+	}
+
+	changeset := dag.LoadChangesetFromID(dagger.ChangesetID(changesetID))
+
+	var summary strings.Builder
+	if err := (func() (rerr error) {
+		ctx, span := Tracer().Start(ctx, "analyzing changes")
+		defer telemetry.End(span, func() error { return rerr })
+
+		patch, err := changeset.AsPatch().Contents(ctx)
+		if err != nil {
+			return fmt.Errorf("get patch contents: %w", err)
+		}
+
+		return summarizePatch(idtui.NewOutput(&summary), patch, -1)
+	})(); err != nil {
+		return err
+	}
+
+	exportDest := outputPath
+	if exportDest == "" {
+		exportDest = "."
+
+		var confirm bool
+		form := idtui.NewForm(
+			huh.NewGroup(
+				huh.NewConfirm().
+					Title("Apply changes?").
+					Description(summary.String()).
+					Affirmative("Apply").
+					Negative("Discard").
+					Value(&confirm),
+			),
+		)
+		if err := Frontend.HandleForm(ctx, form); err != nil {
+			return err
+		}
+		if !confirm {
+			return nil
+		}
+	}
+
+	ctx, span := Tracer().Start(ctx, "applying changes")
+	defer telemetry.End(span, func() error { return rerr })
+	slog := slog.SpanLogger(ctx, InstrumentationLibrary)
+
+	if _, err := changeset.Layer().Export(ctx, exportDest); err != nil {
+		return err
+	}
+
+	removed, err := changeset.RemovedPaths(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
Question? If I added `Host.removeFiles` or similar, would this be able to be moved to the engine?

Dunno if it's a huuuge blocker, but I can't see if it's possible to export a changeset from `dagger shell`.